libs/SalesforceSDKCore/SalesforceSDKCore/Classes/NativeLoginManagerInternal.swift

```diff
@@ -215,7 +211,303 @@ public class NativeLoginManagerInternal: NSObject, NativeLoginManager {
+                        recaptchaevent: reCaptchaParameterGenerationResult.enterpriseReCaptchaEvent,
+                        userData: UserData(
+                            email: email,
+                            username: username,
+                            password: newPassword,
+                            firstName: firstName,
+                            lastName: lastName),
+                        otpVerificationMethod: otpVerificationMethodString)
+                ),
+                encoding: .utf8)!
+            } catch let error {
+                SFSDKCoreLogger().e(
+                    classForCoder,
+                    message: "Cannot JSON encode start registration request body due to an encoding error with description '\(error.localizedDescription)'.")
+                return nil
+            }}() else { return StartRegistrationResult(nativeLoginResult: .unknownError) }
+        
+        // Create the start registration request.
+        let startRegistrationRequest = RestRequest(
+            method: .POST,
+            baseURL: loginUrl,
+            path: kSFOAuthEndPointHeadlessInitRegistration,
+            queryParams: nil)
+        startRegistrationRequest.endpoint = ""
+        startRegistrationRequest.requiresAuthentication = false
+        startRegistrationRequest.setCustomRequestBodyString(
+            startRegistrationRequestBodyString,
+            contentType: kHttpPostApplicationJsonContentType
+        )
+        
+        // Submit the start registration request and fetch the response.
+        let startRegistrationResponse = await withCheckedContinuation { continuation in
+            RestClient.sharedGlobal.send(
+                request: startRegistrationRequest
+            ) { result in
+                continuation.resume(returning: result)
+            }
+        }
+        
+        // React to the start registration response.
+        switch startRegistrationResponse {
+            
+        case .success(let startRegistrationResponse):
+            // Decode the start registration response.
+            guard let startRegistrationResponseBody = {
+                do {
+                    return try startRegistrationResponse.asDecodable(type: StartRegistrationResponseBody.self)
+                } catch let error {
+                    SFSDKCoreLogger().e(
+                        classForCoder,
+                        message: "Cannot JSON decode start registration response body due to a decoding error with description '\(error.localizedDescription)'.")
+                    return nil
+                }}() else { return StartRegistrationResult(nativeLoginResult: .unknownError) }
+            return StartRegistrationResult(
+                nativeLoginResult: .success,
+                email: startRegistrationResponseBody.email,
+                requestIdentifier: startRegistrationResponseBody.identifier)
+            
+        case .failure(let error):
```

Comment:
In here and some other places, it looks like we're swallowing a potentially more detailed error and returning it as "unknown", thoughts on throwing the original errors instead of wrapping them?

Reply:
That's a real great question and maybe a conversation starter.  At the onset of our Salesforce Headless Identity API implementations in https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/3676 `NativeLoginResult.unknown` was introduced in way I interpreted as a general catch-all so none of the public methods of `NativeLoginManager` would have to throw.  There was some discussion later on https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/3688 when we tried setting a few of the methods to throw that we later decided to discard and have return one of the `NativeLoginResult` error cases (they're all error except for success).

My long-term recommendation actually would be to have these methods throw and ensure the catch can access the original failure response from the API for their decision.  That lets us avoid having to add detailed `NativeLoginResult` cases, like I just added for invalid email, each time we want to let a caller see or respond to something different.  I don't think native login manager has any client or server side error case it handles with its own internal logic - it always passes the decision on to the caller. Using throw we also don't have to maintain a pass-through type like `NativeLoginResult` for every error case.  We might even be able to eliminate that enum type entirely and have a simple success or fail result for each action.  Using throw we might also be able to pass-through the actual API results to the caller since there are a lot of them and they could change at any time without an MSDK release in which we create a surrogate `NativeLoginResult` to represent them.

I was also thinking about our style for client-side validation since many mobile projects I've worked don't attempt things like email validation on the client - the API takes responsibility for that in a single place so clients don't have to duplicate the logic.  I'm just thinking aloud on this one, but the Headless Identity API docs do validate parameters such as username, password and email.

The only trouble is that's a different pattern from 12.x, so we'd either have an API change or some methods that throw while others use an error enum case.

I'm really flexible to anything we want to do for the moment.  Shall we keep it as-is so it's consistent with the released API, set the net new code up to throw or something else either now or later?

By the by, I confirmed the API is already handling email validation and returning that in a way we could probably throw meaningfully.

<img width="1791" alt="Screenshot 2024-06-18 at 15 02 13" src="https://github.com/forcedotcom/SalesforceMobileSDK-iOS/assets/94090596/2c83f29c-1679-44b2-8c0b-8847e8ca32cd">

Reply:
This might be interesting when compared to Kotlin's pattern since Swift exceptions are checked at compile time.  Though, there're not when called from Objective-C if I recall.  One conversation point in #3688 was that we never wanted to throw since an exception not checked at compile time might not be caught by the caller.  Maybe we solve for that with documentation?

Reply:
Great conversation here, throwing in my 2¢:

I have never liked the idea of intentionally throwing if it has a possibility of happening in production.  I do however, absolutely see the value in passing that server side error message to the client (one way or another).  IMO it would be great if we could attach a dynamic string to the unknown enum value for the app to log of whatever.  I am absolutely open to ideas on how to pass the error.

I do however think there is value in the result enum and do not want to see that go away.  A lot of the value we are adding here with this API is that we have thought through and tested the error cases.  I see returning a defined results enum so that the app can simply use a switch on to handle all cases is a real convenience when all they (probably) care about is the UI/UX.

Reply:
Oh and I think client side validation is just nice to have.  It's pretty trivial to implement and allows the app to give the user instant feedback instead of waiting for the (somewhat slow) API call to come back.  It also allows us to give the app a specific error back without having to parse the response and map it to a result.

Reply:
I like your long term recommendation and what you mentioned about the enum, I think `NativeLoginResult` made sense for the earlier cases but it seems like it's getting more complicated with nested result types and more cases.

I'd lean towards making the net new code throw and then could have separate work to adjust the old if wanted? 

For the exception part-- for a method that throws in Swift, the generated Objective-c version has an error argument instead so I don't think there's any safety issue there (apps that want the error information would provide one to potentially be populated). Here's an example of the original vs generated version and [usage](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/a8d90a6be90897319513dbfc1a3d0abcef859640/libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFSDKEncryptedURLCache.m#L91-L92)

```
public static func encrypt(data: Data, key: Data?) throws -> Data

+ (NSData * _Nullable)encryptData:(NSData * _Nonnull)data key:(NSData * _Nullable)key error:(NSError * _Nullable * _Nullable)error
```

Reply:
It sounds like we might have a plan as I paraphrase through the notes above and try to find the best common ground.  I'll try to summarize a direction, but am still open to letting others choose since everything on the table "works" for the customer (if I were the customer, that's what I'd say) and we're just choosing style.

1. We can't attach any values to this enum, unfortunately, since it's an Objective-C enum rather than Swift.  It only gets to be an int in the end.  
2. Thanks for the clarification regarding throwing Swift functions and how they handle in Objective-C.
3. Do we have a majority consensus to introduce throws on the new methods here when the API call fails and ensure that the API provided error is available to the caller? If we do, I can push a commit with that.  If we don't I believe this pull request merges as-is.

Option "C" is we can revisit this topic later as well, which might be good so we can brown bag on our error styles in our own time without incrementing our cycle time on this issue.

Thanks for the ideas and help in moving this one to the goal 🚀

Reply:
I kind of like option "C" so we could pull in the remainder of the team to discuss our error styles on our own time.

Reply:
I'm good with throwing or option C, @brandonpage what do you think?

Reply:
@JohnsonEricAtSalesforce @bbirman I was going to suggest something along the lines of "option C".  I think we leave it as is for now.  If we want to remove the return value/type and throw (which I am still not a huge fan of) we should do it across the board -- which would be breaking.  So we might as well lump that in with the name changes Eric wanted to make to password-less login APIs in a major release.  We can discuss as a team what we want the API to be for iOS and Android.   

The timeline I propose:
- Leave this for now.
- Deprecate in 12.1 or 12.2.  
- Implement new API in 12.2 or 13.0.
- Remove existing API in 13.0